app/src/main/java/protect/card_locker/importexport/StocardImporter.java

```diff
@@ -121,173 +161,180 @@ public ZIPData importZIP(ZipInputStream zipInputStream, final ZIPData zipData) t
-                    );
+                    record.frontImage = ZipUtils.readImage(zipInputStream);
                 } else if (fileName.endsWith("/images/back.png") || fileName.endsWith("/images/back/back.jpg")) {
-                    loyaltyCardHashMap = appendToHashMap(
-                            loyaltyCardHashMap,
-                            cardName,
-                            "backImage",
-                            ZipUtils.readImage(zipInputStream)
-                    );
+                    record.backImage = ZipUtils.readImage(zipInputStream);
                 }
+            } else {
+                Log.d(TAG, "Unknown or unused file " + fileName + ", skipping...");
             }
         }
 
-        return new ZIPData(loyaltyCardHashMap, providers);
+        return new ZIPData(cards, providers);
     }
 
-    public ImportedData importLoyaltyCardHashMap(Context context, final ZIPData zipData) {
+    public ImportedData importLoyaltyCardHashMap(Context context, final ZIPData zipData) throws FormatException {
         ImportedData importedData = new ImportedData(new ArrayList<>(), new HashMap<>());
         int tempID = 0;
 
-        for (Map<String, Object> loyaltyCardData : zipData.loyaltyCardHashMap.values()) {
-            String providerId = (String) loyaltyCardData.get("_providerId");
+        List<String> cardKeys = new ArrayList<>(zipData.cards.keySet());
+        Collections.sort(cardKeys);
 
-            if (providerId == null) {
-                Log.d(TAG, "Missing providerId for card " + loyaltyCardData + ", ignoring...");
+        for (String key : cardKeys) {
+            StocardRecord record = zipData.cards.get(key);
+
+            if (record.providerId == null) {
+                Log.d(TAG, "Missing providerId for card " + record + ", ignoring...");
                 continue;
             }
 
-            HashMap<String, Object> providerData = zipData.providers.get(providerId);
+            if (record.cardId == null) {
+                throw new FormatException("No card ID listed, but is required");
+            }
+
+            StocardProvider provider = zipData.providers.get(record.providerId);
 
             // Read store from card, if not available (old export), fall back to providerData
-            String store;
-            if (loyaltyCardData.containsKey("store")) {
-                store = (String) loyaltyCardData.get("store");
-            } else {
-                store = providerData != null ? providerData.get("name").toString() : providerId;
+            String store = record.store != null ? record.store : provider != null ? provider.name : record.providerId;
+            String note = record.note != null ? record.note : "";
+            String barcodeTypeString = record.barcodeType != null ? record.barcodeType : provider != null ? provider.barcodeFormat : null;
+
+            if (record.label != null && !record.label.equals(store) && !record.label.equals(note)) {
```

Comment:
@TheLastProject This seemed like a good way to handle the labels. But I'm not sure what they look like in practice.

Reply:
Actually, looking at the results of a grep of files given to me the label is most of the time an empty string (`"label": ""`) but when it is defined it tends to be a value like "Gold" or "Platinum". So it is probably not a store name but better used in the Note field.

Sorry for getting it wrong at first.

Reply:
No problem. Will update.

Reply:
Seems good to me now.

Reply:
Thanks for checking how it's actually used :)